user-creation/src/main/java/no/unit/nva/useraccessservice/usercreation/person/cristin/CristinPersonRegistry.java

```diff
@@ -336,13 +347,29 @@ private <T> T executeRequest(HttpRequest request, Class<T> type) {
     }
 
     private void assertSuccessResponse(HttpRequest request, HttpResponse<String> response) {
-        if (response.statusCode() < 200 || response.statusCode() >= 300) {
+        int statusCode = response.statusCode();
+        if (statusCode / ONE_HUNDRED != SUCCESS_FAMILY) {
             var message = generateErrorMessageForResponse(request, response);
             LOGGER.info(message);
-            throw new PersonRegistryException(message);
+
+            if (HTTP_NOT_FOUND == statusCode) {
+                throw new IdentityServiceNotFoundException(message);
+            } else if (isAlreadyExists(statusCode, message)) {
+                throw new IdentityServiceAlreadyExistsException(message);
+            } else if (HTTP_INTERNAL_ERROR <= statusCode) {
+                throw new IdentityServiceUpstreamInternalServerErrorException(message);
+            } else {
+                throw new IdentityServiceException(IdentityServiceErrorCodes.GENERIC_ERROR, message);
+            }
         }
     }
 
+    private static boolean isAlreadyExists(int statusCode, String message) {
+        return HTTP_CONFLICT == statusCode
+               || (HTTP_BAD_REQUEST == statusCode
+                   && message.contains("already exists"));
```

Comment:
at the moment this is the best way to detect this, but added future support for 409 conflict

Reply:
Lol. This reminds me of systems that returned 200 + XML for errors. Oh, let's not talk about "modern protocols" :D